### PR TITLE
feat: allow query params outside of the hashbang

### PR DESCRIPTION
**Pull request by anxolin**
# Summary

Normal links should include the query params after the hashbang, however, right know there's some links out there with this shape:

`https://swap.cow.fi/swap?param1=value1&param2=value2`

Good ones:
`https://swap.cow.fi/#/swap?param1=value1&param2=value2`

This PR just handles the case we "forget" to add the hash bang. Although we still should, to not cause too many redirects

# Context
https://cowservices.slack.com/archives/C036G0J90BU/p1691507761164199

# To Test

TEST

1. https://swap-dev-git-allow-query-params-out-of-hashbang-cowswap.vercel.app
2. https://swap-dev-git-allow-query-params-out-of-hashbang-cowswap.vercel.app/
3. https://swap-dev-git-allow-query-params-out-of-hashbang-cowswap.vercel.app/?
4. https://swap-dev-git-allow-query-params-out-of-hashbang-cowswap.vercel.app/?utm_source=github
5. https://swap-dev-git-allow-query-params-out-of-hashbang-cowswap.vercel.app?utm_source=github
6. https://swap-dev-git-allow-query-params-out-of-hashbang-cowswap.vercel.app?other=parameter
7. https://swap-dev-git-allow-query-params-out-of-hashbang-cowswap.vercel.app?utm_source=github&other=parameter
8. https://swap-dev-git-allow-query-params-out-of-hashbang-cowswap.vercel.app/#/
8. https://swap-dev-git-allow-query-params-out-of-hashbang-cowswap.vercel.app/#/swap
9. https://swap-dev-git-allow-query-params-out-of-hashbang-cowswap.vercel.app/#/swap?utm_source=github
10. https://swap-dev-git-allow-query-params-out-of-hashbang-cowswap.vercel.app/#/swap?utm_source=github&other=parameter
11. https://swap-dev-git-allow-query-params-out-of-hashbang-cowswap.vercel.app/#/5/swap/WETH?utm_source=github
12. https://swap-dev-git-allow-query-params-out-of-hashbang-cowswap.vercel.app/#/5/swap/WETH?utm_source=github&other=parameter

### CODE REVIEW

File: src/modules/utm/hooks.ts [hunk 1]
```diff
@@ -65,7 +66,12 @@ export function useInitializeUtm(): void {
       // Clear params from URL and redirect
       const cleanedParams = cleanUpParams(searchParams)
       if (cleanedParams) {
-        navigate({ pathname, search: searchParams.toString() }, { replace: true })
+        const newSearch = searchParams.toString()
+        if (hasQueryParamsOutOfHashbang) {
+          window.location.replace(newSearch ? `${pathname}?${newSearch}` : '/')
```
Comment (alfetopito) [inline src/modules/utm/hooks.ts:71]: Shouldn't it always replace the location with at least `pathname` to remove the original params?

I don't think it makes sense to keep the trash in the url like this example: 

Test #6
From: https://swap-dev-git-allow-query-params-out-of-hashbang-cowswap.vercel.app?other=parameter
To: https://swap-dev-git-allow-query-params-out-of-hashbang-cowswap.vercel.app/?other=parameter#/1/swap/WETH

**Review by alfetopito — COMMENTED** [review]
On Test steps #10 and #11 (https://swap-dev-git-allow-query-params-out-of-hashbang-cowswap.vercel.app/#/1/swap/WETH?utm_source=github), the parameter is not consumed when the app loads the first time.
It I try to load it again then it is removed

**Comment by elena-zh** [comment]
Hey @anxolin , 
I have connected #189 issue to this PT as it fixes it. 
The same a Leandro mentions, params are still displayed to me in cases 6-7 

and 10, 11, 13



Also, when I remove # from the link when edit it manually, the link becomes expanded, invalid parameters are saved in it. Is this OK?
type in

**Comment by shoom3301** [comment]
@elena-zh @alfetopito I've addressed issues above.
One case:
`http://localhost:3000/100/limit/WXDAI?other=parameter&utm_content=ccc&tab=history` should redirect to `http://localhost:3000/#/100/limit/WXDAI?other=parameter&page=1&tab=history`.

How it works:
1. We move the path `/100/limit/WXDAI` after `#`
2. We remove `utm_content` for the query
3. We keep `other=parameter&tab=history`
4. Limit orders widget adds `page=1` to the query

**Comment by elena-zh** [comment]
Hey @shoom3301 , I removed # from the link and got this: 
https://swap-dev-git-allow-query-params-out-of-hashbang-cowswap.vercel.app/1/limit/WETH?page=1&tab=open
image

Also, cases 11, 13: params are not removed from the link:

**Comment by shoom3301** [comment]
@elena-zh thanks! Fixed

**Comment by alfetopito** [comment]
Test 3: `?` remains in the url
  `https://swap-dev-git-allow-query-params-out-of-hashbang-cowswap.vercel.app?` -> `https://swap-dev-git-allow-query-params-out-of-hashbang-cowswap.vercel.app/?#/1/swap/WETH`
  
Test 11 and 13: parameter remains in the URL. Expected?
  `https://swap-dev-git-allow-query-params-out-of-hashbang-cowswap.vercel.app/#/swap?utm_source=github&other=parameter` -> `https://swap-dev-git-allow-query-params-out-of-hashbang-cowswap.vercel.app/#/1/swap/WETH?other=parameter`

**Comment by shoom3301** [comment]
>Test 11 and 13: parameter remains in the URL. Expected?

Yep, this is expected. We can't know which parameters are useful in the app, so we can't delete them

**Comment by shoom3301** [comment]
>Test 3: ? remains in the url

@alfetopito fixed it: https://github.com/cowprotocol/cowswap/pull/3025/commits/574522230b078fa5ad40a6fe449968a3a89b7011

I'm not sure this is a real case, anyway :)

**Comment by alfetopito** [comment]
> > Test 3: ? remains in the url
> 
> @alfetopito fixed it: [5745222](https://github.com/cowprotocol/cowswap/commit/574522230b078fa5ad40a6fe449968a3a89b7011)
> 
> I'm not sure this is a real case, anyway :)

Try it with some other garbage for example:

https://swap-dev-git-allow-query-params-out-of-hashbang-cowswap.vercel.app?akhsdias -> `https://swap-dev-git-allow-query-params-out-of-hashbang-cowswap.vercel.app/#/?akhsdias`

I think the additional garbage might be ok (I rather remove it but whatever), but the rest of the app path is not there.

File: apps/cowswap-frontend/src/modules/utm/hooks.ts [hunk 2]
```diff
@@ -53,20 +47,36 @@ export function useInitializeUtm(): void {
...
-        navigate({ pathname, search: searchParams.toString() }, { replace: true })
+      const newSearch = cleanUpParams(searchParams).toString()
+
+      if (hasQueryParamsOutOfHashbang) {
+        window.location.replace(newSearch ? `/#${locationPath}?${newSearch}` : '/')
+        return
       }
+
+      const validHref = `${origin}${locationPath}${hash}${locationSearch}`
+      const isWeirdURl = href !== validHref
```
Comment (anxolin) [inline apps/cowswap-frontend/src/modules/utm/hooks.ts:71]: LoL!

Comment (anxolin) [reply apps/cowswap-frontend/src/modules/utm/hooks.ts:71]: nit: isNotCanonical? isNotStardardUrl?

**Review by anxolin — COMMENTED** [review]
APPROVE! works great with all URLs

**Review by elena-zh — APPROVED** [review]
Looks good now!

**Pull request merged**